### PR TITLE
Add a read only runtime host

### PR DESCRIPTION
**Pull request by expenses**

### CODE REVIEW

File: src/executor/read_only_runtime_host.rs [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+                    self.vm = tx.resume();
+                }
+
+                host::HostVm::EndStorageTransaction { resume, rollback } => {
+                    if rollback {
+                        todo!() // TODO:
+                    }
+
+                    self.vm = resume.resume();
+                }
```
Comment (tomaka) [inline src/executor/read_only_runtime_host.rs:365]:
```suggestion
```

This is all forbidden.

File: src/executor/read_only_runtime_host.rs [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+                    self.vm = req.into();
+                    return RuntimeHostVm::NextKey(NextKey {
+                        inner: self,
+                        key_overwrite: None,
+                    });
+                }
+
+                host::HostVm::ExternalOffchainStorageSet(req) => {
+                    self.vm = req.resume();
+                }
```
Comment (tomaka) [inline src/executor/read_only_runtime_host.rs:321]:
```suggestion
```

Forbidden

File: src/executor/read_only_runtime_host.rs [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+                        self.vm = req.into();
+                        return RuntimeHostVm::StorageGet(StorageGet { inner: self });
+                    }
+                }
+
+                host::HostVm::ExternalStorageSet(req) => {
+                    self.top_trie_changes
+                        .insert(req.key().to_vec(), req.value().map(|v| v.to_vec()));
+                    self.vm = req.resume();
+                }
```
Comment (tomaka) [inline src/executor/read_only_runtime_host.rs:309]:
```suggestion
```

Verboten

File: src/executor/read_only_runtime_host.rs [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+}
+
+/// Implementation detail of the execution. Shared by all the variants of [`RuntimeHostVm`]
+/// other than [`RuntimeHostVm::Finished`].
+struct Inner {
+    /// Virtual machine running the call.
+    vm: host::HostVm,
+
+    /// Pending changes to the top storage trie that this execution performs.
+    top_trie_changes: HashMap<Vec<u8>, Option<Vec<u8>>, fnv::FnvBuildHasher>,
```
Comment (tomaka) [inline src/executor/read_only_runtime_host.rs:269]:
```suggestion
```

This simplifies a lot of the rest of the code. I'm not going to add suggestions everywhere.

File: src/executor/read_only_runtime_host.rs [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+}
+
+/// Fetching the key that follows a given one is required in order to continue.
+#[must_use]
+pub struct NextKey {
+    inner: Inner,
+
+    /// If `Some`, ask for the key inside of this field rather than the one of `inner`. Used in
+    /// corner-case situations where the key provided by the user has been erased from storage.
+    key_overwrite: Option<Vec<u8>>,
```
Comment (tomaka) [inline src/executor/read_only_runtime_host.rs:173]:
```suggestion
```

Shouldn't be needed anymore

File: src/executor/read_only_runtime_host.rs [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+    .run())
+}
+
+/// Execution is successful.
+#[derive(Debug)]
+pub struct Success {
+    /// Contains the output value of the runtime, and the virtual machine that was passed at
+    /// initialization.
+    pub virtual_machine: SuccessVirtualMachine,
+    pub logs: String,
```
Comment (tomaka) [inline src/executor/read_only_runtime_host.rs:62]: Missing docs

File: src/executor/read_only_runtime_host.rs [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+                    if self.logs.len().saturating_add(message.len()) >= 1024 * 1024 {
+                        return RuntimeHostVm::Finished(Err(Error::LogsTooLong));
+                    }
+
+                    self.logs.push_str(&message);
+                    self.vm = req.resume();
+                }
+
+                _ => {
+                    panic!("Forbidden");
```
Comment (tomaka) [inline src/executor/read_only_runtime_host.rs:384]:
```suggestion
                    RuntimeHostVm::Finished(Err(Error::ForbiddenHostCall))
```

(add the `ForbiddenHostCall` variant to `Error`)

**Review by tomaka — CHANGES_REQUESTED** [review]

**Review by tomaka — APPROVED** [review]
:+1:

**Comment by tomaka** [comment]
You left as a draft, but I suppose that this is ready?

**Comment by expenses** [comment]
Yep

**Comment by tomaka** [comment]
Testing the new `automerge` feature of #413!

**Pull request merged**